pydsdl/_dsdl.py

```diff
@@ -0,0 +1,259 @@
+        """
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def file_path(self) -> Path:
+        """The path to the DSDL file on the filesystem."""
+        raise NotImplementedError()
+
+    @property
+    @abstractmethod
+    def root_namespace_path(self) -> Path:
+        """
+        The path to the root namespace directory on the filesystem.
+        """
+        raise NotImplementedError()
+
+
+class DsdlFileBuildable(DsdlFile):
+    """
+    A DSDL file that can construct a composite type from its contents.
+    """
+
+    @abstractmethod
+    def read(
+        self,
+        lookup_definitions: Iterable["DsdlFileBuildable"],
+        definition_visitors: Iterable["DefinitionVisitor"],
+        print_output_handler: Callable[[int, str], None],
+        allow_unregulated_fixed_port_id: bool,
+    ) -> CompositeType:
+        """
+        Reads the data type definition and returns its high-level data type representation.
+        The output should be cached; all following invocations should read from this cache.
+        Caching is very important, because it is expected that the same definition may be referred to multiple
+        times (e.g., for composition or when accessing external constants). Re-processing a definition every time
+        it is accessed would be a huge waste of time.
+        Note, however, that this may lead to unexpected complications if one is attempting to re-read a definition
+        with different inputs (e.g., different lookup paths) expecting to get a different result: caching would
+        get in the way. That issue is easy to avoid by creating a new instance of the object.
+        :param lookup_definitions:              List of definitions available for referring to.
+        :param definition_visitors:             Visitors to notify about discovered dependencies.
+        :param print_output_handler:            Used for @print and for diagnostics: (line_number, text) -> None.
+        :param allow_unregulated_fixed_port_id: Do not complain about fixed unregulated port IDs.
+        :return: The data type representation.
+        """
+        raise NotImplementedError()
+
+
+DefinitionVisitor = Callable[[DsdlFile, DsdlFileBuildable], None]
+"""
+Called by the parser after if finds a dependent type but before it parses a file in a lookup namespace.
+:param DsdlFile argument 0: The target DSDL file that has dependencies the parser is searching for.
+:param DsdlFile argument 1: The dependency of target_dsdl_file file the parser is about to parse.
+"""
+
+SortedFileT = TypeVar("SortedFileT", DsdlFile, DsdlFileBuildable, CompositeType)
+SortedFileList = List[SortedFileT]
+"""A list of DSDL files sorted by name, newest version first."""
+
+FileSortKey: Callable[[SortedFileT], Tuple[str, int, int]] = lambda d: (
+    d.full_name,
+    -d.version.major,
+    -d.version.minor,
+)
```

Comment:
Why lambda when you could use a well-typed function?

```suggestion
def get_definition_ordering_rank(d: DSDLFile | CompositeType) -> tuple[str, int, int]:
    return d.full_name, -d.version.major, -d.version.minor
```